File: cost/iona/orbacus/java/main/TestIntfContextDSI_impl.java
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package ORBTest;

import org.omg.CORBA.*;

class TestIntfContextDSI_impl
    extends org.omg.PortableServer.DynamicImplementation
{
    private ORB m_orb;
    private ORBTest_Context.Intf m_ti;

    TestIntfContextDSI_impl(ORB orb, ORBTest_Context.Intf ti)
    {
	m_orb = orb;
	m_ti = ti;
    }

    static final String[] m_ids =
    {
        "IDL:ORBTest_Context/Intf:1.0"
    };

    public String[]
    _all_interfaces(
	org.omg.PortableServer.POA poa,
	byte[] object_id
    )
    {
        return m_ids;
    }

    public void
    invoke(
	ServerRequest request
    )
    {
	String name = request.operation();

	boolean ex;
	if(name.length() > 2 && name.endsWith("Ex"))
	{
	    name = name.substring(0, name.length() - 2);
	    ex = true;
	}
	else
	    ex = false;
	
	if(name.equals("opContext"))
	{
	    NVList list = m_orb.create_list(0);
	    Any any = m_orb.create_any();
	    any.type(m_orb.get_primitive_tc(TCKind.tk_string));
	    list.add_value("", any, org.omg.CORBA.ARG_IN.value);
	    request.arguments(list);
	    org.omg.CORBA.Context context = request.ctx();
	    String pattern = any.extract_string();

            try
            {
                String[] ret = m_ti.opContext(pattern, context);

                Any result = m_orb.create_any();
		ORBTest_Context.StringSequenceHelper.insert(result, ret);
                request.set_result(result);
            }
            catch(BAD_CONTEXT e)
            {
                Any anyEx = m_orb.create_any();
                BAD_CONTEXTHelper.insert(anyEx, e);
                request.set_exception(anyEx);
            }
	
	    return;
	}

	System.err.println("DSI implementation: unknown operation: " + name);

        NVList list = m_orb.create_list(0);
        request.arguments(list);

        Any exAny = m_orb.create_any();
        BAD_OPERATIONHelper.insert(exAny, new BAD_OPERATION());
        request.set_exception(exAny);
    }
}

File: cost/tests/orbacus/main/TestIntfContextDSI_impl.cpp
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <TestIntfContextDSI_impl.h>

using namespace CORBA;

TestIntfContextDSI_impl::TestIntfContextDSI_impl(
    ORB_ptr orb,
    ORBTest_Context::Intf_ptr ti
)
    : m_orb(ORB::_duplicate(orb)),
      m_ti(ORBTest_Context::Intf::_duplicate(ti))
{
}

RepositoryId
TestIntfContextDSI_impl::_primary_interface(
    const PortableServer::ObjectId& oid,
    PortableServer::POA_ptr poa
)
{
    return string_dup("IDL:ORBTest_Context/Intf:1.0");
}

CORBA::Boolean
TestIntfContextDSI_impl::_is_a(
    const char* name
)
{
    if (strcmp(name, "IDL:ORBTest_Context/Intf:1.0") == 0)
    {
	return true;
    }
    else
    {
	return PortableServer::ServantBase::_is_a(name);
    }
}

void
TestIntfContextDSI_impl::invoke(
    ServerRequest_ptr request
)
{
    String_var name = request->operation();

    bool ex;
    if (
	strlen(name) > 2
	&& strcmp((const char*)name + strlen(name) - 2, "Ex") == 0
    )
    {
	name[strlen(name) - 2] = '\0';
	ex = true;
    }
    else
    {
	ex = false;
    }

    if (strcmp(name, "opContext") == 0)
    {
	NVList_ptr list;
	Any* pany;
        Any any;

	m_orb->create_list(0, list);
	list->add(ARG_IN)->value()->replace(_tc_string, 0);
	request->arguments(list);
	Context_ptr ctx = request->ctx();

	pany = list->item(0)->value();
	const char* pattern;
	*pany >>= pattern;

	try
	{
	    ORBTest_Context::StringSequence* ret =
		m_ti->opContext(pattern, ctx);

	    any <<= ret;
	    request->set_result(any);
	}
	catch (const BAD_CONTEXT& ex)
	{
	    any <<= ex;
	    request->set_exception(any);
	}

	return;
    }
}
